proc/threads.c

```diff
@@ -924,8 +924,7 @@ int proc_threadPriority(int priority)
 
 	current = _proc_current();
 	if (priority >= 0) {
-		/* Don't lower current thread priority if it holds any locks */
-		if ((priority < current->priority) || (current->locks == NULL)) {
+		if (priority < _proc_threadGetPriority(current)) {
```

Comment:
Hmm, `_proc_threadGetPriority` cannot return priority lower than `current->priority` (because `current-priority` is either `current->priorityBase` or it has been set by previous call to `_proc_threadGetPriority)`, so if `priority < _proc_threadGetPriority(current)` then also `priority < current->priority`. Or am I missing something?

Reply:
You're right, the current thread priority shouldn't be taken into account here at all. Only priority of the threads waiting on the locks taken by the current thread matter. I've added new function `_proc_threadGetLockPriority()` which returns thread `lock priority` - max priority of the threads waiting for any of the locks taken by the given thread.